enhancements/network/allow-mtu-changes.md

```diff
@@ -0,0 +1,207 @@
+changes in the underlay or because they were set incorrectly at install time.
+
+### Goals
+
+* Allow to change MTU post install on OVN Kubernetes.
+
+### Non goals
+
+* Change the MTU without service disruption.
+
+* Other safe or unsafe configuration changes.
+
+## Proposal
+
+The CNO monitors changes on the operator configuration. When it detects a MTU
+change:
+1. Set the `clusteroperator/network` conditions:
+   - Progressing: true
+   - Upgradeable: false
+2. Deploy pods on every node with `restartPolicy: Never` which are responsible
+   for validating the preconditions. If the preconditions are met the pod will
+   exit with code 0. Some of the preconditions that we will check are:
+   - For MTU changes, the new MTU is valid to apply on that node.
+3. Once all the previous pods are finished, deploy other set of pods on every
+   node that will handle the actual change of the MTU. Wait for them to be
+   ready and running.
+4. Cordon every node, we don't want pods created from this point on.
+5. Ensure via machine-config-operator, that upon reboot, configure-ovs will
+   set the br-ex and its physical port MTUs to the new MTU value.
+6. Ensure that the configmap ovnkube-config is synchronized with the new MTU
+   value.
+7. Set the new MTU value to the applied cluster configuration. At this point
+   CNO is commited to the MTU change.
+8. If any step was unsuccesful, the CNO will set the `clusteroperator/network`
+   conditions to:
+   - Progressing: false
+   - Degraded: true
+   Update the operator configuration status with a description of the problem.
+   At this point the process is interrupted and we require manual intervention.
+9. Wait for pods of step 3 to complete.
+10. Force a rollout of the ovnkube-master and ovnkube-node daemonset sequentially
+    and in appropriate order. This will ensure ovn-kubernetes uses the new MTU
+    value as well.
+11. If any of the previous steps failed, reboot the node, wait for the kubelet to
+    be reporting as Ready again and uncordon it. Otherwise just uncordon the node.
+12. Upon completion, set conditions to:
+    - Progressing: false
+    - Degraded: false
+
+The steps to change the MTU performed by pods of previous step 3 are:
+1. So that we don't have nodes doing things at different times and we have
+   everything synchronized, the pods will wait until the MTU value on the
+   applied cluster configuration changes.
+2. Enter the network namespace of every pod in the pod network and change
+   the MTU of `eth0`. At the same time, the pod will change the MTU of the
+   end of the veth pair in the virtual switch.
+3. The pod will change the MTU of the interfaces ovn-k8s-mp0, ovn-k8s-gw0
+   (if in local mode) and br-int sequentially in that order without any delay
+   in between.
+4. The pod will change the MTU of br-ex and it's physical port sequentially
+   in that order without any delay in between. This includes changing the
+   corresponding NM connection profiles.
```

Comment:
The CNO only allows configuring the _SDN MTU_, not the node MTU. In particular, if you set the CNO config's MTU option at install time, we don't change the node MTUs. So if we change the node MTUs when you change the CNO config post-install-time, then that means the option has a different meaning depending on when you set it, which is no good.

Previously we had said that, assuming the administrator even wants to change the node MTUs (which they may not), then they are responsible for doing it themselves, either before the sdn MTU change if increasing MTU, or after the sdn MTU change if decreasing MTU.

Reply:
Ok, I will consider then the node MTU a separate action that needs to be triggered manually. And what about adding a new setting to CNO: `node-mtu`. On omission it would the current node MTU.

I guess that currently there is a check that the SDN MTU, if specified,  has a valid value according to the node MTU?

Reply:
Yes, openshift-sdn and ovn-kubernetes will refuse to start if the requested tunnel MTU is too large relative to the node MTU.